biz.aQute.bnd.api/src/biz/aQute/bnd/facade/api/Binder.java

```diff
@@ -0,0 +1,304 @@
+		assert prev == null;
+		cycles++;
+	}
+
+	void release() {
+		synchronized (lock) {
+			this.current = null;
+			close(registration.getAndSet(null));
+			lock.notifyAll();
+			cycles++;
+		}
+	}
+
+	@Override
+	public void close() {
+		release();
+		synchronized (lock) {
+			if (closed)
+				return;
+			closed = true;
+		}
+		synchronized (binders) {
+			binders.remove(this);
+		}
+	}
+
+	/**
+	 * Check if this Binder's facade has been gc'ed.
+	 */
+
+	public boolean hasGone() {
+		return facade.get() == null;
+	}
+
+	/**
+	 * Get the backing service object. Since it is possible that there is not
+	 * yet such an object, this will wait a {@link #timeout} number of
+	 * nanoseconds before giving up.
+	 */
+	@Override
+	public D get() {
+
+		synchronized (lock) {
+			if (closed)
+				throw new IllegalStateException("the binder with id " + id + " has been closed for " + description);
+
+			if (current != null)
+				return current;
+
+			try {
+				long start = System.nanoTime();
+
+				while (current == null) {
+					if ((System.nanoTime() - start) > timeout) {
+						throw new IllegalStateException(
+								"no backing service " + id + " can be found for " + description);
+					}
+					lock.wait(1_000);
+					if (closed)
+						throw new IllegalStateException(
+								"the binder with id " + id + " has been closed for " + description);
+				}
+				return current;
+			} catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				throw new IllegalStateException(
+						"interrupted while waiting for backing service " + id + " for " + description);
+			}
+		}
+	}
+
+	/**
+	 * Provide the backing service without waiting
+	 */
+	public D peek() {
+		synchronized (lock) {
+			return current;
+		}
+	}
+
+	/**
+	 * Set the backing service to a new value
+	 */
+	@Override
+	public void accept(D t) {
+		synchronized (lock) {
+			if (!closed)
+				this.current = t;
+			lock.notifyAll();
+		}
+	}
+
+	/**
+	 * Return the ID of the backing service
+	 */
+	public String getId() {
+		return id;
+	}
+
+	/**
+	 * Get a description of the facade object
+	 */
+	public String getDescription() {
+		return description;
+	}
+
+	/**
+	 * Set the timeout to wait for the backing service to be injected.
+	 *
+	 * @param timeout
+	 *                    timeout in milliseconds
+	 */
+	public void setTimeout(long timeout) {
+		this.timeout = TimeUnit.MILLISECONDS.toNanos(timeout);
+	}
+
+	/**
+	 * Get the saved state of this Binder
+	 */
+
+	public Object getState() {
+		return state;
+	}
+
+	/**
+	 * Get the the state to save
+	 */
+	public void setState(Object state) {
+		this.state = state;
+
+	}
+
+	@Override
+	public String toString() {
+		return "Binder[id=" + id + "]";
+	}
+
+	private static void close(AutoCloseable c) {
+		if (c != null) {
+			try {
+				c.close();
+			} catch (Exception e) {
+				// ignore
+			}
+		}
+	}
+
+	/*
+	 * This adapter is only for test purposes
+	 */
+	public interface TestAdapterNotApiWillChange {
```

Comment:
As a hardware engineer I am familiar with special hardware only for testing but I agree it is not common in software, unfortunately. 

The problem I needed something here is that it is API. Binder is in an API package with lots of implementation code. This is uncommon and not desired but it is the implication of statics. Since my testing code is in the package of the facade or facade manager, the Binder package is not accessible via package access.

However, I moved the interface to the API package and that seems to work.